ocaml/idl/json_backend/gen_json.ml

```diff
@@ -541,31 +541,46 @@ let write_to_dir dir maybe_file =
   Option.iter write maybe_file
 
 module Version = struct
-  type t = int list
-
-  let rec compare x y =
-    match (x, y) with
-    | x :: xs, y :: ys when x = y ->
-        compare xs ys
-    | x :: _, y :: _ ->
-        Int.compare x y
-    | x :: _, [] when x = 0 ->
-        0
-    | _ :: _, [] ->
-        1
-    | [], y :: _ when y = 0 ->
-        0
-    | [], _ :: _ ->
-        -1
-    | [], [] ->
-        0
+  (* versions ending in -next are considered unreleased *)
+  type t = int list * string option
+
+  (* orders like this:
+     - 22.10.0
+     - 22.11.0
+     - 22.11.0-next
+     - 22.12.0
+     From the way the lifecycle binary generates versions there only will be a
+     single -next version, with the highest version number
+  *)
+  let compare (xs, x_next) (ys, y_next) =
+    let cmp = List.compare Int.compare xs ys in
+    if cmp = 0 then
+      Option.compare String.compare x_next y_next
+    else
+      cmp
+
+  let of_name name =
+    let to_version v = String.split_on_char '.' v |> List.map int_of_string in
```

Comment:
This could be also done with `Astring.String.fields`

```
utop # let sep = function '.' | '-' -> true | _ -> false;;
val sep : char -> bool = <fun>
─( 15:51:55 )─< command 2 >──────────────────────────────────────────{ counter: 0 }─
utop # Astring.String.fields ~is_sep:sep "12.13.14-foo";;
- : string list = ["12"; "13"; "14"; "foo"]
```

Reply:
```
utop # Scanf.sscanf "12.23.45" "%d.%d.%d%s" (fun x y z l -> (x,y,z,l));;
- : int * int * int * string = (12, 23, 45, "")

utop # Scanf.sscanf "12.23.45-foo" "%d.%d.%d%s" (fun x y z l -> (x,y,z,l));;
- : int * int * int * string = (12, 23, 45, "-foo")
```

Reply:
the leading dash is unimportant right, now, I've used the scanf concoction